ES|QL Support for ST_GEOHASH, ST_GEOTILE and ST_GEOHEX

craigtaverner opened the pull request:
In support of the feature request at https://github.com/elastic/elasticsearch/issues/123903

Checklist:

* Prototype
  * [x] Implement basic ST_GEOHASH on points
  * [x] Add support for optional bounds parameter
  * [x] Support multivalues (incoming and outgoing)
  * [x] Add support for ST_GEOTILE
  * [x] Add support for ST_GEOHEX
* Productize:
  * [x] Use bounded predicates to filter by tile, not geometry
    * [x] Use `GeoTileBoundedPredicate` for correct `geotile` bounds filtering
    * [x] Use `GeohashBoundedPredicate` for correct `geohash` bounds filtering
    * [x] Copy `GeoHexCellIdSource.GeoHexPredicate` and `H3SphericalUtil` classes into esql module and use them for correct H3 bounds filtering
  * [x] Refine API
    * [x] Change functions to return `long` grid-ids instead of `keyword` for performance and add new functions for `keyword` alternatives
    * [x] Create conversion functions between `long` and `keyword` versions of the grid-ids
    * [x] Create functions to generate geometries from the grid-ids
  * [x] Additional refinements
    * [x] Add unit tests and generate docs
    * [x] Tests for geo_grid queries (`| WHERE ST_GEOHEX(location, 1) == ST_GEOHEX_TO_LONG("812bbffffffffff")`)
    * [x] Check input parameters for validity (eg. negative precision)
    * [x] Errors tests
    * [x] Precision range validation
    * [x] Remove support for `_TO_GEOSHAPE` functions
    * [x] Remove support for dynamic precision (from field or calculated values)
    * [x] Test coverage of bounds (and get bounds types into the docs)
    * [x] Restrict bounds parameter to only `geo_shape` of an envelope (no hidden envelope calculation)
    * [x] Test that prove docs-values code path works (ie. function before spatial stats)
    * [x] Add license checks for geohex grid
    * [x] Docs links to query dsl
* Extra features (could be done here or in later PRs) 
  * [ ] Bring back support for `_TO_GEOSHAPE` functions, including enhanced WKB (include grid-id)
  * [ ] Lucene pushdown for geo_grid queries (see `GeoGridQueryBuilder`, and copy to ES|QL)
  * [ ] Expand support to `geo_shape` fields (including license checks)
  * [ ] Add pushdown to Lucene for grid aggregations
  * [ ] Bring back support for dynamic precision (from field or calculated values)
  * [ ] Non-foldable bounds

Fixes #123903

craigtaverner committed:
Add initial geohash function to ESQL

craigtaverner committed:
Add generated files

craigtaverner committed:
Add support for optional bounds parameter

craigtaverner committed:
Added geotile and geohex

craigtaverner committed:
Add ESQL Capability and reduce test flakiness

iverase commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/spatial/StGeotile.java:
```diff
@@ -0,0 +1,362 @@
+    }
+
+    protected static BytesRef calculateGeotile(Point point, int precision, Rectangle bounds) {
+        // For points, filtering the point is as good as filtering the tile
```
I don't think this statement is correct. See https://github.com/elastic/elasticsearch/blob/368d47131f2d74742ae89005680352389387bdb8/server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoTileBoundedPredicate.java

craigtaverner replied:
I agree. The current code was a quick-and dirty placeholder for the draft PR. I'll definitely switch over to the better code you suggest.

craigtaverner committed:
Fix failing EsqlNodeSubclassTests from incorrect NodeInfo

craigtaverner committed:
Use GeoTileBoundedPredicate for geotile bounds filtering

craigtaverner committed:
Added ST_GEOXXX_TO_LONG, _TO_STRING and _TO_GEOSHAPE functions

This is because we want to move the ST_GEOXXX functions to produce longs only, for performance reasons, and need the convenience functions to display these either as strings or as actual shapes.

This is both useful to the users, as well as easier to do manual tests where we visualize the results on a map.

craigtaverner committed:
Convert geogrid functions to produce only long values

craigtaverner committed:
Fixes after rebasing on main

iverase commented:
I was trying queries like:

```
ROW location=TO_GEOPOINT(\"POINT(0 0)\")| EVAL hex=ST_GEOHEX(location, -1)"
```

In case of geohex and geotile we provide a good error although looking into the stack trace it feels to me that we are not checking the input while parsing and we only fail when trying to apply the function to the point. More over, for geohash it does not throw an error.

Let's improve the checking of input parameters.

craigtaverner committed:
Support geo-bounded-predicate for geohash and geohex

This involved copying the H3 utils from spatial plugin to esql. We should rather find a common location or dependency for these.

craigtaverner committed:
Added unit tests for new functions and resulting docs

craigtaverner committed:
Output POLYGON instead of BBOX for TO_GEOSHAPE functions

craigtaverner committed:
Make test less flaky

craigtaverner committed:
Use wrapper classes on the bounded predicates to simplify and reduce duplicated code.

craigtaverner committed:
Spotless

craigtaverner committed:
Add precision validation as well as more examples, refined docs, and support for geo_point bounds

craigtaverner committed:
Support depthOffset in MD docs headings for nesting functions

craigtaverner committed:
Add depthOffset to geo grid functions

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into esql_grid_aggs

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into esql_grid_aggs

craigtaverner committed:
Added type error tests and reduced precision to only support integer

craigtaverner committed:
Added geo_grid queries (no lucene pushdown)

craigtaverner committed:
Added tests for invalid precision

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

craigtaverner committed:
Update docs/changelog/125143.yaml

elasticsearchmachine commented:
Hi @craigtaverner, I've created a changelog YAML for you.

craigtaverner committed:
Fix docs links

craigtaverner committed:
Merge branch 'main' into esql_grid_aggs

craigtaverner committed:
Merge branch 'esql_grid_aggs' of github.com:craigtaverner/elasticsearch into esql_grid_aggs

elasticsearchmachine commented:
Hi @craigtaverner, I've updated the changelog YAML for you.

craigtaverner committed:
Update docs/changelog/125143.yaml

craigtaverner committed:
Fix invalid changelog

craigtaverner committed:
Fix function count

craigtaverner committed:
Merge branch 'main' into esql_grid_aggs

craigtaverner committed:
Some more specific tests to flush out bounds differences on serverless

craigtaverner committed:
Fix new tests for serverless

craigtaverner committed:
Another test for serverless

craigtaverner committed:
Check if the issue is related to crossing the dateline

craigtaverner committed:
Remove support for grid TO_GEOSHAPE functions

iverase commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/spatial-grid.csv-spec:
```diff
@@ -0,0 +1,938 @@
+required_capability: spatial_grid
+
+FROM airports
+| WHERE ST_GEOHEX(location, 1) == ST_GEOHEX_TO_LONG("812bbffffffffff")
```
Can we have a test for geo_shape as well? I expect it does not work for that as it is an equality on multivalues.

craigtaverner replied:
I listed geo_shape as something to support in a followup PR. I did implemented the evaluators in a mv-friendly way, so it would be easier to support geo_shape, but right now this feature is geo_point only.

iverase commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/spatial/StGeotile.java:
```diff
@@ -0,0 +1,292 @@
+        int precision,
+        @Fixed GeoBoundingBox bbox
+    ) {
+        GeoTileBoundedGrid bounds = new GeoTileBoundedGrid(precision, bbox);
```
Are we creating a GeoTileBoundedGrid per document here?

craigtaverner replied:
This particular code path is for when the precision comes from the document, so yes, we are creating the predicate per document. I tried to make sure that all cases where we know stuff earlier, we create the predicate earlier. I noticed that not all predicates actually use the precision internally, but GeoTileBoundedPredicate does.

For literal precision, see line 166 where the GeoTileBoundedGrid is created for the entire execution.

iverase replied:
>This particular code path is for when the precision comes from the document,

iverase replied:
what's the use case this is trying to cover?

craigtaverner committed:
Delete docs files for removed functions

craigtaverner committed:
More fixes to docs after removing _to_GEOSHAPE functions

craigtaverner committed:
Merge branch 'main' into esql_grid_aggs

craigtaverner committed:
Fix functions count after removing to_geoshape methods

craigtaverner committed:
Remove all field evaluators except spatial field evaluator

And refine tests to better cover bounds parameter, fixing missing type information  in the docs

iverase commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/spatial-grid.csv-spec:
```diff
@@ -0,0 +1,887 @@
+FROM airports
+| EVAL points = ["POINT(0.0 30.0)", "POINT(12.0 60.0)"]
+| EVAL geohash = ST_GEOHASH(location, 2, TO_GEOPOINT(points))
+| WHERE geohash IS NOT NULL
```
Why is this needed?

iverase replied:
I think I understand now. It is strange how the bounds are defined here.

craigtaverner replied:
Support for bounds defined by points is now removed.

craigtaverner committed:
Remove support for geo_point defined bounds

Now only geo_shape bounds are supported

craigtaverner committed:
Disallow any geo_shape bounds other than BBOX

Also enhance tests to validate against actual bounds

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into esql_grid_aggs

craigtaverner committed:
Add license checks and integration tests to test license checks

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into esql_grid_aggs

iverase reviewed (approved):
LGTM

I think we need to strength the test coverage for multi-value case but we can do it in a follow up

craigtaverner committed:
Fix license checks in function unit tests

craigtaverner committed:
Experiment to reduce serverless flakiness

craigtaverner committed:
Merge branch 'main' into esql_grid_aggs

craigtaverner committed:
Merge branch 'main' into esql_grid_aggs

elasticsearchmachine commented:
## 💔 Backport failed
| Status | Branch | Result |
|:------:|:------:|:------:|
| ❌ |  8.19  | Commit could not be cherrypicked due to conflicts |

You can use [sqren/backport](https://github.com/sqren/backport) to manually backport by running `backport --upstream elastic/elasticsearch --pr 125143`
